### PR TITLE
feat(gatsby-plugin-google-tagmanager): GDPR consent + default options

### PR DESCRIPTION
**Pull request by crstnio**
## Description

As of the GDPR it is a best practice to gather the user agreement on setting cookies and/or enabling tracking scripts. It's the developers responsibility to enable tracking/cookies on user consent and to remove cookies if the user declines afterwards.

But, we should give the developers options to make that possible and their jobs easier.

This PR allows two more options:

- `gdpr` to disable Google Tag Manager on page load. Set to `false` by default.
- `gdprConsent` for a global variable (inside `window`) that holds the users consent. Results in `window['gdprConsent']` by default.

`gdpr` could be set by an environment variable, that is set depending on the users location (EU or not) and controlled by the ADN.

Again, it is the developers responsibility to gather the consent and set `window[gdprConsent]`. This value should come from a cookie and I plan to develop a plugin/component that enables the user to agree, sets both the cookie and variable, executes optional functions and renders blocked children.

The code is now wrapped inside a `gtmCode()` function that is needed to enable Google Tag Manager without reloading the page.

### Object destructuring

This feature introduces default options by destructuring the options object. It seemed appropriate to add this with the GDPR feature.

The default value of `includeInDevelopment` (README) is now set too.

## Related Issues

@sebastienfi [uses GA](https://github.com/gatsbyjs/gatsby/issues/4976#issuecomment-381344919) (comment) in a similar way, but with a `window.gdpr-agreed` variable.

### CODE REVIEW

**Comment by wardpeet** [comment]
Thanks for creating this PR, I'm just not sure this is a good fix for gdpr & googletagmanager.

Googletag manager doesn't break GDPR regulations. It's what googletag is configured to do, that might have issues with GPDR. Adding these gdpr variables in my opinion isn't a correct fix and might make this plugin a bit more complicated to use.

The correct fix is to actually push something on your DataLayer that tells gtm to load or not load scripts.

If this is hard to do I'm ok with creating a shouldLoad option that has a function to skip loading which is more general approach than just checking a global variable.

What do you think? (also want to involve @gatsby/core into this discussion)

**Comment by cardiv** [comment]
Hm, I tried the approach of using `dataLayer` events that trigger tags in GTM.

To control this you would need to add a script (for browser) + the `dataLayer` (ssr) before GTM is loaded. Time for another plugin or build this right into `gatsby-plugin-google-tagmanager`?

**Comment by wardpeet** [comment]
Yes, that's true. I just don't think `GDPR` should be part of GTM. What do you think of creating a shouldLoad option where you can pass in a function?

**Comment by cardiv** [comment]
The `shouldLoad` option would only be evaluated at build time, right?

Since the user needs to agree in the browser, this would be no solution I think.

As you said, `dataLayer` seems the way to go. But that requires the ability to push before GTM both at build time and in the browser.

We could add an option that takes a function which will be injected before the GTM code and allows for early `dataLayer` actions. Maybe the GTM code could be blocked until `dataLayer` is ready.

**Comment by wardpeet** [comment]
I'll make 2 examples one with dataLayer and one with the shouldLoad function. I was thinking a function that's ran inside the browser.

**Comment by wardpeet** [comment]
@cardiv would something like this work?

File: packages/gatsby-plugin-google-tagmanager/README.md [hunk 1]
```diff
@@ -12,14 +12,19 @@ Easily add Google Tagmanager to your Gatsby site.
...
+    resolve: "gatsby-plugin-google-tagmanager",
     options: {
       id: "YOUR_GOOGLE_TAGMANAGER_ID",
 
       // Include GTM in development.
       // Defaults to false meaning GTM will only be loaded in production.
       includeInDevelopment: false,
 
+      // datalayer to be set before GTM is loaded
+      // should be a stringified object or object
```
Comment (cardiv) [inline packages/gatsby-plugin-google-tagmanager/README.md]: `generateDefaultDataLayer` checks for a function too. So not only objects are supported, right?

File: packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js [hunk 2]
```diff
@@ -1,46 +1,70 @@
...
+  })(window,document,'script','dataLayer', '${id}');`
+
+const generateGTMIframe = ({ id, environmentParamStr }) =>
+  oneLine`<iframe src="https://www.googletagmanager.com/ns.html?id=${id}${environmentParamStr}" height="0" width="0" style="display: none; visibility: hidden"></iframe>`
+
+const generateDefaultDataLayer = (dataLayer, reporter) => {
+  let result = `window.dataLayer = window.dataLayer || [];`
+
+  if (typeof dataLayer === `function`) {
+    result += `window.dataLayer.push((${dataLayer})()));`
```
Comment (cardiv) [inline packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js]: Aren't there way too much brackets at the end? Or am I missing something here?

**Comment by cardiv** [comment]
Nice work @wardpeet, thank you! I am a bit confused about two code parts but this could handle some more use cases.

File: packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js (hunk 2 above)
Comment (wardpeet) [reply packages/gatsby-plugin-google-tagmanager/src/gatsby-ssr.js]: good catch 1 to many 😄

File: packages/gatsby-plugin-google-tagmanager/README.md (hunk 1 above)
Comment (wardpeet) [reply packages/gatsby-plugin-google-tagmanager/README.md]: thanks will have to update this comment, it indeed allows a function and an object :) stringified object isn't possible anymore.

**Comment by wardpeet** [comment]
@cardiv could you test if it works now. I updated the test so we catch my errors 👍

**Comment by wardpeet** [comment]
@cardiv ping :) I would love to get this merged.

**Comment by freiksenet** [comment]
There is some cleanup that needs to be done before merging. @wardpeet will address this as soon as he has time. Thank you for your patience!

**Pull request merged**